components/raftstore/src/store/local_metrics.rs

```diff
@@ -123,317 +62,62 @@ impl RaftSendMessageMetrics {
-            self.read_index_no_leader = 0;
-        }
-        if self.region_not_initialized > 0 {
-            RAFT_INVALID_PROPOSAL_COUNTER
-                .region_not_initialized
-                .inc_by(self.region_not_initialized);
-            self.region_not_initialized = 0;
-        }
-        if self.is_applying_snapshot > 0 {
-            RAFT_INVALID_PROPOSAL_COUNTER
-                .is_applying_snapshot
-                .inc_by(self.is_applying_snapshot);
-            self.is_applying_snapshot = 0;
-        }
-    }
-}
-
-#[derive(Clone, Default)]
-pub struct RaftLogGcSkippedMetrics {
-    pub reserve_log: u64,
-    pub threshold_limit: u64,
-    pub compact_idx_too_small: u64,
-}
-
-impl RaftLogGcSkippedMetrics {
-    fn flush(&mut self) {
-        if self.reserve_log > 0 {
-            RAFT_LOG_GC_SKIPPED.reserve_log.inc_by(self.reserve_log);
-            self.reserve_log = 0;
-        }
-        if self.threshold_limit > 0 {
-            RAFT_LOG_GC_SKIPPED
-                .threshold_limit
-                .inc_by(self.threshold_limit);
-            self.threshold_limit = 0;
-        }
-        if self.compact_idx_too_small > 0 {
-            RAFT_LOG_GC_SKIPPED
-                .compact_idx_too_small
-                .inc_by(self.compact_idx_too_small);
-            self.compact_idx_too_small = 0;
-        }
+    pub fn flush(&mut self) {
+        self.0.flush();
     }
 }
 
 /// The buffered metrics counters for raft.
-#[derive(Clone)]
 pub struct RaftMetrics {
-    pub store_time: LocalHistogram,
-    pub ready: RaftReadyMetrics,
+    // local counter
+    pub ready: RaftReadyCounterVec,
     pub send_message: RaftSendMessageMetrics,
-    pub message_dropped: RaftMessageDropMetrics,
-    pub propose: RaftProposeMetrics,
+    pub message_dropped: RaftDroppedMessageCounterVec,
+    pub propose: RaftProposalCounterVec,
+    pub invalid_proposal: RaftInvalidProposalCounterVec,
+    pub raft_log_gc_skipped: RaftLogGcSkippedCounterVec,
+
+    // local histogram
+    pub store_time: LocalHistogram,
+    pub propose_wait_time: LocalHistogram,
     pub process_ready: LocalHistogram,
     pub commit_log: LocalHistogram,
-    pub leader_missing: Arc<Mutex<HashSet<u64>>>,
-    pub invalid_proposal: RaftInvalidProposeMetrics,
     pub write_block_wait: LocalHistogram,
+
+    // waterfall metrics
     pub waterfall_metrics: bool,
     pub wf_batch_wait: LocalHistogram,
     pub wf_send_to_queue: LocalHistogram,
     pub wf_send_proposal: LocalHistogram,
     pub wf_persist_log: LocalHistogram,
     pub wf_commit_log: LocalHistogram,
     pub wf_commit_not_persist_log: LocalHistogram,
-    pub raft_log_gc_skipped: RaftLogGcSkippedMetrics,
+
+    pub leader_missing: Arc<Mutex<HashSet<u64>>>,
+
+    last_flush_time: Instant,
 }
 
 impl RaftMetrics {
     pub fn new(waterfall_metrics: bool) -> Self {
         Self {
+            ready: RaftReadyCounterVec::from(&STORE_RAFT_READY_COUNTER_VEC),
```

Comment:
@BusyJay here is the key change. Before we define `RaftReadyMetrics` manually by repeating the fields in the static metrics definition. Now we use the local of the static metrics directly.

Reply:
it's quite like calling `STORE_RAFT_READY_COUNTER_VEC.local()` but a static version

Reply:
see https://github.com/tikv/rust-prometheus/blob/master/static-metric/examples/simple.rs#L24

Reply:
Yes, my question is what's the difference between using provided APIs and defining it manually? Are they the exact same struct?

Reply:
Yes, it's just a wrapper for `LocalIntCounter` inherently, see generated code https://github.com/tikv/rust-prometheus/blob/master/static-metric/examples/make_auto_flush_static_counter.rs#L270